sn_interface/src/network_knowledge/sections_dag.rs

```diff
@@ -0,0 +1,888 @@
+    }
+}
+
+impl SectionsDAG {
+    /// Creates a new DAG containing only the `genesis_key`
+    pub fn new(genesis_key: bls::PublicKey) -> Self {
+        Self {
+            genesis_key,
+            dag: MerkleReg::new(),
+            dag_root: BTreeSet::new(),
+            hashes: BTreeMap::new(),
+        }
+    }
+
+    /// Insert new key into the DAG. `parent_key` must exist in the DAG and must validate
+    /// `signature`, otherwise error is returned.
+    pub fn insert(
+        &mut self,
+        parent_key: &bls::PublicKey,
+        key: bls::PublicKey,
+        signature: bls::Signature,
+    ) -> Result<()> {
+        if !self.verify_sig(parent_key, &key, &signature) {
+            return Err(Error::InvalidSignature);
+        }
+
+        let parent = match self.hashes.get(parent_key) {
+            Some(hash) => BTreeSet::from([*hash]),
+            None => {
+                // parent set is empty if we are genesis's child
+                if *parent_key == self.genesis_key {
+                    self.dag_root.insert(key);
+                    BTreeSet::new()
+                } else {
+                    return Err(Error::KeyNotFound(*parent_key));
+                }
+            }
+        };
+
+        let info = SectionInfo {
+            key,
+            sig: signature,
+        };
+        // children argument (DAG terminology) is the actual parent
+        let child = self.dag.write(info, parent);
+        self.dag.apply(child.clone());
+        self.hashes.insert(key, child.hash());
+
+        Ok(())
+    }
+
+    /// Returns the list of the leaf sections. A section is considered a leaf if it has not
+    /// gone through any churn or split. Empty list if we only hold the genesis.
+    pub fn leaf_sections(&self) -> Vec<SectionInfo> {
+        self.dag.read().values().cloned().collect()
+    }
+
+    /// Returns the list of the leaf keys
+    pub fn leaf_keys(&self) -> Vec<bls::PublicKey> {
+        // if leaf_sections is empty, genesis is the only leaf
+        let leaf_sections = self.leaf_sections();
+        if leaf_sections.is_empty() {
+            vec![self.genesis_key]
+        } else {
+            leaf_sections.iter().map(|info| info.key).collect()
+        }
+    }
+
+    /// Returns the list of all the keys in the SectionDAG
+    pub fn keys(&self) -> Vec<bls::PublicKey> {
+        // use DFS
+        let mut keys = Vec::new();
+        let mut need_to_visit = vec![self.genesis_key];
+
+        while let Some(key) = need_to_visit.pop() {
+            keys.push(key);
+            let children = match self.get_child_keys(&key) {
+                Ok(children) => children,
+                // unreachable since keys are obtained from self
+                Err(_) => {
+                    error!(
+                        "SectionsDAG::keys() Unreachable since key is obtained from self Can be error only if hashes go out of sync"
+                    );
+                    return Vec::new();
```

Comment:
This is a good candidate for a panic, by returning a vec, we're masking a bug.

Either we panic, or we need to return a Result

```suggestion
                    panic!("SectionsDAG::keys() Unreachable since key is obtained from self Can be error only if hashes go out of sync");
```